README
In this version three classes are used:
drvMBus     This is a standalone class that does all modbus IO.  An instance of
            this class is accessed via a pointer.
drvBkhAsyn  This class handles communication with basic Beckhoff systems.
            It can be used to build an IOC for use with most bBeckhoff hardware
            excluding the stepper motor controllers.
drvBkhAMot  This class is based on drvBkhAsyn class and is intended for
            controlling stepper motors via the Beckhoff KL2531 and KL2541
            bus terminals.

Presently, the low priority message queue is using max number of messages
of 500 and the High priority message is using 300.  All initial reading of
register values is done via the high priority message queue.  Each motor
stepper control driver object is initializing 31 hidden registers.  This
results in 3x or 93 IOs per driver object.  The 300 value may need to be
increased for a larger setup.

Last modified on 7/29/2013 (zms)

In Feb of 2015, rewrite the motor class to add homing functions.  This also
required modifications to consolidate all motor parameters into one object
of the class.  Some small modifications were also needed in the Beckhoff base
class.

Known problems:
1. When a stage is homed and the home button is pressed again before the stage
  is moved, this needs to be aborted by using the Stop button.  To put it
  differently, homing can't be done if the stage is at home position and the
  stage was not yet moved since it was homed.
2. When the stage is moved to the limit thus activating a limit switch, the
  motion is stopped.  Of the three buttons to start motion only the relative
  move button to move the stage away from the limit is active.  The problem
  is that the Go button should also be enabled if the absolute move setup in
  input widget would result in a move away from the limit.  Presently this
  is not working.
3. When the IOC is restarted, it does not know if the stage is homed or not
  eventhough the Beckhoff controller still has the correct home position
  defined.  The home position is defined as 0 position.  This information
  is lost if the Beckhoff controller is power cycled.


RELEASE_NOTES
1/20/2015, zms
In this release the handling of "at limit" situation is reworked.  Previously,
when a stage motion was stopped as a limit swith was activated, the operator
needed to press a push button to move the stage away from the limit switch.
In this release changes are made such that the operator can move the stage
away from the activated limit switch by pressing the push buttons used for
normal motion, while the buttons used to move the stage towards the activated
limit switch remain disabled until the motion stops.

2/25/2014, zms
In this release more code is added to allow the home position be defined to be
either at negative limit switch, positive limit switch, or a current position.
More code is added to home the stage according to definition of home.
If home is defined at a limit switch, then the motor is driven to the limit
switch, when the limit switch is activated, the motion is reversed and the
stage moved slowly until the limit switch is deactivated and this position is
marked as home.
If home is defined as a current position, then the current position is marked
as home.
If the stage is not equipped with limit switches, then only home at current
position is allowed.
A function callable from st.cmd file
  drvBkhAMotSetup( int hometype,int neglim,int poslim)
where hometype can be either
      0 at negative limit switch,
      1 at current position,
      2 at positive limit switch.
      neglim if 1, have a negative limit switch, if 0 do not have one,
      poslim if 1, have a positive limit switch, if 0 do not have one.
Negative limit switch is reached by driving the motor in the direction of
decreasing step count direction, and the positive limit switch is reached by
driving the motor in the direction of increasing step count direction.

Note that one can connect motor coil wires to the controller to get the motor
rotate in the desired direction.  To put it differently, by adjusting the
direction of motor rotation, either of the two limit switches could be defined
either positive or negative.

3/2/2015, zms
In this release functionality was added to home the stage.  A new function
drvBkhAMotSetup( const char* port,int hometype,int neglim,int poslim)
was added, which is callable from st.cmd file.  It is used to define homing
parameters:
hometype: 1 home at negative switch, 2 make home at current
            position, 3 home at positive switch.
neglim if 1 limit switch present at negative position.
poslim if 1 limit switch present at positive position.
If either neglim or poslim is 0, that limit switch is not present.
Implementing the homing functions required some code reorganization.  This
is reflected in how the motor driver is configured in the st.cmd file.
In this release, one object is created for each axis.  Previously, three
objects were created.


bkhApp/mbusSrc/drvMBus.h
/* drvMBus.h
 * Asyn driver class that implements modbus protocol.  It uses the essence
 * from modbus support module written by Mark Rivers.
 * This driver was developed to be used in Beckhoff control IOC.  It does the
 * modbus protocol IO via the doModbusIO routine in modbusAsyn.c.  The IOs are
 * dispatched from an IOThread.  Requests are received from Beckhoff device
 * driver layer via the mbusDoIO method, which are scheduled for execution by
 * placing them in an EPICS message queue.  Entries are removed from the
 * message queue in IOThread and processed in the order received.
 *
 * This device driver does not know anything about Beckhoff, so the requests
 * must be properly phrased.  This is no longer true.  I have implemented
 * a handful of "special" methods which implement a sequence of IO operations
 * needed to achieve complex tasks which need to be done for a single end, like
 * read from or write to a Beckhoff hidden register.  Some of these "special"
 * methods perform functions specific to the KL2531 (KL2541) stepper motor
 * controller bus terminals.  Thus with these functions, complex operations
 * are performed with a single request.
 * Started on 6/26/2013, zms
 *---------------------------------------------------------------------------*/
#ifndef _drvMBus_h
#define _drvMBus_h
#include <epicsMessageQueue.h>
#include <epicsMutex.h>
#include <shareLib.h>
#include "asynDriver.h"
#include "modbus.h"
#include "modbusAsyn.h"

#ifndef SIZE
#define SIZE(x)         (sizeof(x)/sizeof(x[0]))
#endif
#ifndef MIN
#define MIN(a,b)        (((a)<(b))?(a):(b))
#endif
#ifndef MAX
#define MAX(a,b)        (((a)>(b))?(a):(b))
#endif

typedef unsigned char   byte;
typedef epicsUInt16	word;
typedef unsigned int    uint;

#define PLEN	16
#define NMSGQL	400
#define NMSGQH	50
#define NDATA	128

typedef struct{
  void*	pdrv;
  int	addr,chan,a,rn,pix,func;
  word	data[NDATA];
  int	len;
  int	stat;
} iodone_t;
typedef void (*iocb_t)( iodone_t);

typedef struct{
  char port[PLEN],name[PLEN];
  int slave,addr,len;
  modbusDataType_t dt;
} drvd_t;

typedef struct{
  int	maddr;
  int	func;
  int	addr,chan,a,rn,pix;
  int	d,len;
  int	six;		// special index of type spix_t
  void*	pdrv;
} msgq_t;

typedef enum{prioH_e,prioL_e} prio_t;
typedef enum{normal_e,spix0_e,spix1_e,spix2_e,spix3_e,spix4_e,spix5_e} spix_t;

class drvMBus{
public:
  drvMBus( drvd_t dd,int msec);

  void		IOThread();
  asynStatus	mbusDoIO( prio_t prio,int six,int saddr,int addr,int chan,
			int n,int a,int rn,int pix,int func,int len,
			int d,void* pdrv);
  void          exitHndl();
  void          mbusLock();
  void          mbusUnlock();
  void		mbusPurgeQueue( prio_t ix);
  void		registerCB( iocb_t cb);
  void		report();
  void		doHist( int en);
  void		clearHist();
  epicsInt32*	getHist();
  int		getAllowInLQ(){return(_allowInLQ);}
  void		putAllowInLQ( int n){ _allowInLQ=n;}
 
protected:
  void		_emptyQueue( epicsMessageQueue* pmq);
  asynStatus	mbusMemIO( msgq_t msgq);
  void		_doSpecial0( msgq_t msgq);
  void		_doSpecial1( msgq_t msgq);
  void		_doSpecial2( msgq_t msgq);
  void		_doSpecial3( msgq_t msgq);
  void		_doSpecial4( msgq_t msgq);
  void		_doSpecial5( msgq_t msgq);
  int		_readSpecialOne( msgq_t msgq,int r,epicsUInt16* v);
  int		_writeSpecialOne( msgq_t msgq,int r,epicsUInt16 v);
  void		_readSpecial( msgq_t msgq,int r1,int r2);
  epicsMessageQueue* _pmqL;     // low priority message queue
  epicsMessageQueue* _pmqH;     // high priority message queue
  mBus_t*	_pmb;		// pointer to the modbus data structure.

private:
  epicsMutexId  _mutexId;       // mutex for interlocking access to IO
  int		_halt;		// if true disallow IO
  iocb_t	_cb;		// IO done callback
  double	_tout;
  int		_inLQ;		// number of items in low priority queue
  int		_inHQ;		// number of items in high priority queue
  int		_maxInLQ;	// max number in low prio queue
  int		_maxInHQ;	// max number in high prio queue
  int		_allowInLQ;	// max number allowed in low prio queue
  int		_npurgLQ;	// number of times low prio queue was purged
  int		_npurgHQ;	// number of times hi prio queue was purged.
};

epicsShareExtern drvMBus* pmbus=0;

#endif // _drvMBus_h


bkhApp/bkhErrSrc/drvBkhErr.h
/* drvBkhErr.h
 * This driver class, derived from asynPortDriver, is designed to handle error
 * messages and error state generated by all instances of the drvBkhAsyn or
 * drvBkhAMot classes.  Object of this class is created via a C configure
 * function called from the st.cmd file of the IOC and an extern address
 * pointer is made available to the other class objects.
 * Started on 1/9/2015, zms
 *---------------------------------------------------------------------------*/
#ifndef _drvBkhErr_h
#define _drvBkhErr_h
#include <epicsMutex.h>
#include <ellLib.h>
#include "asynPortDriver.h"

#ifndef MIN
#define MIN(a,b)        (((a)<(b))?(a):(b))
#endif
#ifndef MAX
#define MAX(a,b)        (((a)>(b))?(a):(b))
#endif

typedef unsigned char   byte;
typedef epicsUInt16     word;
typedef unsigned int    uint;
typedef struct{
  ELLNODE       node;
  int           eiId,flag;
  char          port[16];
} errItem_t;

#define NSLOTS		1
#define WFLEN		100

#define wfMessageStr    "WF_MESSAGE"
#define siNameStr	"SI_NAME"
#define biErrorStr	"BI_ERROR"
#define boTestStr	"BO_TEST"

class drvBkhErr: public asynPortDriver{
public:
  drvBkhErr( const char* port);

  virtual asynStatus readInt32( asynUser* pau,epicsInt32* v);
  virtual asynStatus writeInt32( asynUser* pau,epicsInt32 v);
  virtual void report( FILE* fp,int level);
  void		exitHndl();
  void		initDone( int flg);
  int		registerClient( const char* port);
  void		setErrorFlag( int id,int flag);
  void		errLock();
  void		errUnlock();

protected:
  void		_message( const char*);
  void		_refresh();

  int	_wfMessage, _siName, _biError, _boTest;

#define FIRST_ITEM _wfMessage
#define LAST_ITEM  _boTest
#define PARMS (&LAST_ITEM - &FIRST_ITEM + 1)

enum{	ixWfMessage, ixSiName, ixBiError, ixBoTest};

private:
  epicsMutexId	_mutexId;
  errItem_t*	_find(int id);
  char*		_port;
  int		_firstix;
  int		_initdone;
  ELLLIST	_clist;		// list of client data structures
  int		_cid;		// client identifier
  char		_msg[WFLEN];	// for error message
};

epicsShareExtern drvBkhErr* pbkherr=0;

#endif // _drvBkhErr_h


bkhApp/mbusSrc/drvMBus.cpp
/* drvMBus.cpp
 * This device driver is a self contained class that does modbus protocol.
 * Started on 6/26/2013, zms.
 *---------------------------------------------------------------------------*/
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include <cantProceed.h>
#include <epicsTypes.h>
#include <epicsString.h>
#include <epicsThread.h>
#include <epicsMutex.h>
#include <epicsTime.h>
#include <errlog.h>
#include <epicsExit.h>
#include <envDefs.h>
#include <epicsExport.h>
#include <iocsh.h>

#include "drvMBus.h"

#define STRLEN	128

static const char *dname="drvMBus";

static char* __getTime(){
/*------------------------------------------------------------------------------
 * Returns a pointer to an ASCI coded string of the current date time.
 *----------------------------------------------------------------------------*/
  epicsTimeStamp etm; static char tms[64];
  struct tm tt; long year; ulong ns;
  epicsTimeGetCurrent( &etm);
  epicsTimeToTM( &tt,&ns,&etm);
  year=tt.tm_year+1900;
  sprintf( tms,"%d/%02d/%02d %02d:%02d:%02d",year,tt.tm_mon+1,
        tt.tm_mday,tt.tm_hour,tt.tm_min,tt.tm_sec);
  return(tms);
}

extern "C"{
static void exitHndlC( void* pvt){
  drvMBus* pthis=(drvMBus*)pvt;
  pthis->exitHndl();
}
static void IOThreadC( void* p){
  drvMBus* pthis=(drvMBus*)p;
  pthis->IOThread();
}
}

void drvMBus::IOThread(){
/*-----------------------------------------------------------------------------
 * Process tasks, one at a time.
 *---------------------------------------------------------------------------*/
  static const char* iam="IOThread"; int inq,stat; asynStatus st;
  msgq_t msgq;
printf( "%s::%s:=== tout=%f ======\n",dname,iam,_tout);
  while(1){
    stat=_pmqH->tryReceive( &msgq,sizeof(msgq));
    if(stat==-1){
      inq=_pmqL->pending();
      stat=_pmqL->tryReceive( &msgq,sizeof(msgq));
    }
    if(stat==-1){ epicsThreadSleep(_tout);}
    else{
      st=mbusMemIO( msgq);
      if(st){
	printf( "%s::IOThread:=-=-= six=%d,maddr=%d,pix=%d\n",
		dname,msgq.six,msgq.maddr,msgq.pix);
	fflush(0);
      }
    }
  }
}
void drvMBus::exitHndl(){
/*-----------------------------------------------------------------------------
 *---------------------------------------------------------------------------*/
  _halt=1;
  errlogPrintf( "%s::%s:exitHndl: Clear ID\n",dname,_pmb->name);
}
void drvMBus::mbusLock(){
/*-----------------------------------------------------------------------------
 * Lock access to resource.
 *---------------------------------------------------------------------------*/
  epicsMutexMustLock( _mutexId);
}
void drvMBus::mbusUnlock(){
/*-----------------------------------------------------------------------------
 * Unlock access to resource.
 *---------------------------------------------------------------------------*/
  epicsMutexUnlock( _mutexId);
}
void drvMBus::mbusPurgeQueue( prio_t ix){
/*-----------------------------------------------------------------------------
 * empties the specified queue of all entries.
 *---------------------------------------------------------------------------*/
  epicsMessageQueue* pmq;
  if(ix==prioH_e) pmq=_pmqH; else pmq=_pmqL;
  mbusLock();
  _emptyQueue( pmq);
  mbusUnlock();
}
void drvMBus::_emptyQueue( epicsMessageQueue* pmq){
/*-----------------------------------------------------------------------------
 * empty pmq queue.
 *---------------------------------------------------------------------------*/
  int i,j,stat; msgq_t msgq;
  i=j=pmq->pending();
  while(i){
    stat=pmq->tryReceive( &msgq,sizeof(msgq));
    if(stat==-1) break;
    i=pmq->pending();
  }
printf( "%s::emptyQ: j=%d, i=%d\n",dname,j,i);
}
void drvMBus::report(){
/*-----------------------------------------------------------------------------
 * Prints a report to the console.
 *---------------------------------------------------------------------------*/
  printf( "Report for driver %s --- %s ---\n",dname,__getTime());
  printf( "   %d in Low  prio queue (%d max)\n",_inLQ,_maxInLQ);
  printf( "   %d in High prio queue (%d max)\n",_inHQ,_maxInHQ);
  printf( "   number of purges: LowPQ = %d, HiPQ = %d\n",_npurgLQ,_npurgHQ);
}
asynStatus drvMBus::mbusDoIO( prio_t prio,int six,int saddr,int addr,int chan,
		int n,int a,int rn,int pix,int func,int len,int d,void* pdrv){
/*-----------------------------------------------------------------------------
 * Schedule an IO request by passing it to the IOThread via message que.
 * prio is either prioL_e low priority or prioH_e for high priority,
 * six is a special IO flag, when it is equal normal_e do normal processing,
 * saddr is modbus starting address,
 * addr is the parameter library address index for the pix parameter index,
 * chan is the channel number of a bus terminal, e.g. 32 channel ADC,
 * n and a are used to calculate the modbus address maddr=saddr+n*addr+a,
 * rn is a register number needed for some special functions,
 * pix is an index of a parameter in a parameter library,
 * func is a modbus function for this IO,
 * d is data to be written, it is not used for read requests,
 * len is the number of words of data to be either read or written,
 * pdrv requesting driver object address.
 *---------------------------------------------------------------------------*/
  epicsMessageQueue* pmq; static msgq_t msgq; int j,k,stat;
  if(prio==prioH_e) pmq=_pmqH; else pmq=_pmqL;
  msgq.maddr=saddr+n*chan+a; msgq.six=six;
  msgq.addr=addr; msgq.a=a; msgq.rn=rn; msgq.pix=pix;
  msgq.func=func; msgq.d=d;
  msgq.len=len; msgq.pdrv=pdrv;
  _inLQ=j=_pmqL->pending();
  _inHQ=k=_pmqH->pending();
  if(j>_maxInLQ) _maxInLQ=j;
  if(k>_maxInHQ) _maxInHQ=k;
  if((prio==prioL_e)&&(j>=_allowInLQ)){
    _emptyQueue( pmq);
    _npurgLQ++;
printf( "%s::mbusDoIO: prio=%d,inq=%d,allow=%d,purg=%d\n",
dname,prio,j,_allowInLQ,_npurgLQ);
  }
  stat=pmq->trySend( &msgq,sizeof(msgq));
  if(stat==-1){
    _emptyQueue( pmq);
    if(prio==prioH_e) _npurgHQ++; else _npurgLQ++;
    stat=pmq->trySend( &msgq,sizeof(msgq));
printf( "%s::mbusDoIO: prio=%d,inqL=%d,allow=%d,inqH=%d,purg=%d,%d\n",
dname,prio,j,_allowInLQ,k,_npurgLQ,_npurgHQ);
  }
  return(asynSuccess);
}
asynStatus drvMBus::mbusMemIO( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Called from IOThread.  Does IO on a process image memory block.
 * msgq is the message received from a requester.
 *---------------------------------------------------------------------------*/
  static iodone_t iodone;
  int st=0; epicsUInt16* pw=(epicsUInt16*)iodone.data;
  if(_halt) return(asynSuccess);
  if(!_pmb){
    errlogPrintf( "%s::mbusMemIO, _pmb=%p, bad address\n",dname,_pmb);
    return(asynError);
  }
  switch(msgq.six){
    case spix0_e:	_doSpecial0(msgq); break;
    case spix1_e:	_doSpecial1(msgq); break;
    case spix2_e:	_doSpecial2(msgq); break;
    case spix3_e:	_doSpecial3(msgq); break;
    case spix4_e:	_doSpecial4(msgq); break;
    case spix5_e:	_doSpecial5(msgq); break;
    case normal_e:
      if(msgq.func>MODBUS_READ_INPUT_REGISTERS) *pw=msgq.d;
      st=doModbusIO( _pmb,_pmb->slave,msgq.func,msgq.maddr,pw,msgq.len);
      if(st){
	errlogPrintf( "%s::mbusMemIO:doModbusIO:six=%d failed, st=%d\n",
			dname,msgq.six,st);
      }
      iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
      iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
      iodone.stat=st;
      if(!_cb){
	errlogPrintf( "%s::mbusMemIO: bad callback address\n",dname);
	return(asynError);
      }
if(st){
printf( "%s::mbusMemIO: pdrv=%p,maddr=0x%x,addr=%d,pix=%d,func=%d"
",len=%d,w=(%d,%d)\n",dname,iodone.pdrv,msgq.maddr,msgq.addr,msgq.pix,
msgq.func,msgq.len,pw[0],pw[1]);
fflush(0);
}
      (*_cb)(iodone);
      break;
    default:	return(asynError);
  }
  if(st) return(asynError); else return(asynSuccess);
}
void drvMBus::_doSpecial0( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Read the Beckhoff motor absolute position.
 *---------------------------------------------------------------------------*/
  _readSpecial( msgq,0,1);
}
void drvMBus::_doSpecial1( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Read Beckhoff motor set point (set position) registers.
 *---------------------------------------------------------------------------*/
//printf( "%s::doSpecial1:--------------------\n",dname);
  _readSpecial( msgq,2,3);
}
void drvMBus::_doSpecial2( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Read Beckhoff hidden register.
 *---------------------------------------------------------------------------*/
  static iodone_t iodone; epicsUInt16 w; int st;
  st=_readSpecialOne( msgq,msgq.rn,&w);
  if(!st) iodone.data[0]=w; else iodone.data[0]=0;
//printf( "%s::doSpecial2:---- st=%d,addr=%d,rn=%d,pix=%d,w=0x%x\n",
//dname,st,msgq.addr,msgq.rn,msgq.pix,w);
  iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
  iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
  iodone.stat=st;
  if(!_cb){
    errlogPrintf( "%s::_doSpecial2: bad callback address\n",dname);
  }
  (*_cb)(iodone);
}
void drvMBus::_doSpecial3( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Write to Beckhoff RAM hidden  register.
 *---------------------------------------------------------------------------*/
  static iodone_t iodone; int st;
  st=_writeSpecialOne( msgq,msgq.rn,msgq.d);
//printf( "%s::doSpecial3:---- st=%d ----\n",dname,st);
  iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
  iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
  iodone.stat=st;
  if(!_cb){
    errlogPrintf( "%s::_doSpecial3: bad callback address\n",dname);
  }
  (*_cb)(iodone);
}
void drvMBus::_doSpecial4( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Write to Beckhof hidden EEPROM register.
 *---------------------------------------------------------------------------*/
  static iodone_t iodone;
  epicsUInt16 cbe,w; int st,woff=0x800,maddr;
  int wfunc=MODBUS_WRITE_SINGLE_REGISTER;
  int rfunc=MODBUS_READ_HOLDING_REGISTERS;
  maddr=msgq.maddr+woff;
  st=doModbusIO( _pmb,_pmb->slave,rfunc,maddr,&cbe,1);
  if(!st){
    w=0xc0+31;
    st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
    if(!st){
      w=0x1235;
      st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr+1,&w,1);
      if(!st){
	w=0xc0+msgq.rn;
	st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
	if(!st){
	  w=msgq.d;
	  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr+1,&w,1);
	  if(!st){
	    w=0xc0+31;
	    st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
	    if(!st){
	      w=0;
	      st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr+1,&w,1);
	      if(!st){
		w=0;
		st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
		if(!st){
		  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&cbe,1);
		}
	      }
	    }
	  }
	}
      }
    }
  }
//printf ( "%s::doSpecial4:---- st=%d ----\n",dname,st);
  iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
  iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
  iodone.stat=st;
  if(!_cb){
    errlogPrintf( "%s::_doSpecial4: bad callback address\n",dname);
  }
  (*_cb)(iodone);
}
void drvMBus::_doSpecial5( msgq_t msgq){
/*-----------------------------------------------------------------------------
 * Write 32 bit value to Beckhof hidden set point registers (RAM).
 *---------------------------------------------------------------------------*/
  static iodone_t iodone; int st; epicsUInt16* pw=(epicsUInt16*)&msgq.d;
  int wfunc=MODBUS_WRITE_SINGLE_REGISTER,woff=0x800,maddr;
  epicsUInt16 w=0;
  st=_writeSpecialOne( msgq,msgq.rn,pw[0]);
  if(!st) st=_writeSpecialOne( msgq,msgq.rn+1,pw[1]);
  // next write 0 to data out register, needed for next move
  maddr=msgq.maddr+woff+1;
  if(!st) st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
//printf( "%s::doSpecial5:-- st=%d,d=%d\n",dname,st,msgq.d);
  iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
  iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
  iodone.stat=st;
  if(!_cb){
    errlogPrintf( "%s::_doSpecial5: bad callback address\n",dname);
  }
  (*_cb)(iodone);
}
int drvMBus::_writeSpecialOne( msgq_t msgq,int rn,epicsUInt16 v){
/*-----------------------------------------------------------------------------
 * write to one Beckhoff hidden RAM register.
 *---------------------------------------------------------------------------*/
  epicsUInt16 cbe,w; int st,woff=0x800,maddr;
  int wfunc=MODBUS_WRITE_SINGLE_REGISTER;
  int rfunc=MODBUS_READ_HOLDING_REGISTERS;
  maddr=msgq.maddr+woff;
  st=doModbusIO( _pmb,_pmb->slave,rfunc,maddr,&cbe,1);
  if(st) return(st);
  w=0xc0+rn;		// register number to control byte
  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&w,1);
  if(st) return(st);
  w=v;
  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr+1,&w,1);
  if(st) return(st);
  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&cbe,1);
//printf( "%s::writeSpecialOne: maddr=0x%x,cbe=0x%x,v=%d(0x%x),rn=%d\n",
//dname,maddr,cbe,v,v,rn);
  return(st);
}
void drvMBus::_readSpecial( msgq_t msgq,int r1,int r2){
/*-----------------------------------------------------------------------------
 * Predefined special sequence of modbus IOs.  This is for reading a 32 bit
 * value from two Beckhoff 16 bit hidden registers, register number r1 and r2.
 *---------------------------------------------------------------------------*/
  static iodone_t iodone;
  int st; epicsUInt16 w;

  iodone.data[0]=iodone.data[1]=0;
  st=_readSpecialOne( msgq,r1,&w);
  if(!st){
    iodone.data[0]=w;
    st=_readSpecialOne( msgq,r2,&w);
    if(!st) iodone.data[1]=w;
  }
//printf( "%s::readSpecial: maddr=%d,pix=%d,r1=%d,r2=%d,"
//"d=%d(0x%x),%d(0x%x),st=%d\n",
//dname,msgq.maddr,msgq.pix,r1,r2,iodone.data[0],iodone.data[0],
//iodone.data[1],iodone.data[1],st);
  iodone.addr=msgq.addr; iodone.a=msgq.a; iodone.pix=msgq.pix;
  iodone.func=msgq.func; iodone.len=msgq.len; iodone.pdrv=msgq.pdrv;
  iodone.stat=st;
  if(!_cb){
    errlogPrintf( "%s::_readSpecial: bad callback address\n",dname);
  }
  (*_cb)(iodone);
}
int drvMBus::_readSpecialOne( msgq_t msgq,int r,epicsUInt16* v){
/*-----------------------------------------------------------------------------
 * Read from a Beckhoff hidden register number r and return value in v.
 *---------------------------------------------------------------------------*/
  int wfunc=MODBUS_WRITE_SINGLE_REGISTER; int st;
  int rfunc=MODBUS_READ_HOLDING_REGISTERS;
  int woff=0x800,maddr; epicsUInt16 cb,cbe;
  maddr=msgq.maddr+woff; cb=0x80+r; *v=0;
  st=doModbusIO( _pmb,_pmb->slave,rfunc,maddr,&cbe,1);
  if(st) return(st);
  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&cb,1);
  if(st) return(st);
  st=doModbusIO( _pmb,_pmb->slave,rfunc,msgq.maddr+1,v,1);
  if(st) return(st);
  st=doModbusIO( _pmb,_pmb->slave,wfunc,maddr,&cbe,1);
  return(st);
}
void drvMBus::registerCB( iocb_t cb){
/*--- save callback address for future use ----------------------------------*/
  _cb=cb;
}
void drvMBus::doHist( int en){
/*-----------------------------------------------------------------------------
 * Enable/disable time histogram (en 1/0).
 *---------------------------------------------------------------------------*/
  _pmb->enableHistogram=en;
}
epicsInt32* drvMBus::getHist(){
/*-----------------------------------------------------------------------------
 * Returns a pointer to the time histogram array.
 *---------------------------------------------------------------------------*/
  return(_pmb->timeHistogram);
}
void drvMBus::clearHist(){
/*-----------------------------------------------------------------------------
 * Zeroes the time histogram array.
 *---------------------------------------------------------------------------*/
  memset( _pmb->timeHistogram,0,sizeof(_pmb->timeHistogram));
}
drvMBus::drvMBus( drvd_t dd,int msec){
/*-----------------------------------------------------------------------------
 * Constructor for the drvMBus class. Configures modbus IO routine using data
 * in structure dd.  msec is the IOThread delay in miliseconds.
 *---------------------------------------------------------------------------*/
  mBus_t* pmb=0;
  pmb=modbusAsynConfig( dd.port,dd.slave,3,dd.addr,dd.len,dd.dt,dd.name);

  _pmb=pmb; _halt=0; _cb=0; _tout=msec/1000.0;
  _maxInLQ=_maxInHQ=_npurgLQ=_npurgHQ=0; _allowInLQ=NMSGQL;
  _mutexId=epicsMutexMustCreate();
  _pmqH= new epicsMessageQueue( NMSGQH,sizeof(msgq_t));
  _pmqL= new epicsMessageQueue( NMSGQL,sizeof(msgq_t));

  if(msec){
    epicsThreadCreate( dname,epicsThreadPriorityHigh,
                epicsThreadGetStackSize(epicsThreadStackMedium),
                (EPICSTHREADFUNC)IOThreadC,this);
  }
  epicsAtExit( exitHndlC,this);
  printf( "%s::%s: Port=%s configured\n",dname,dname,dd.port);
}
extern "C" {

int drvMBusConfig( const char* port,int slave,int addr,int len,
		int dtype,const char* name,int msec){
/*-----------------------------------------------------------------------------
 * EPICS iocsh callable function to call constructor for the drvMBus class.
 *  port is the octet port, slave is a modbus slave, addr is modbus starting
 *  address, len is memory length in units of bit or 16 bit word, dtype is
 *  data type (0 if two's complement), name is used in print statements, and
 *  msec is poll routine timeout in mili seconds.
 *---------------------------------------------------------------------------*/
  modbusDataType_t dt=(modbusDataType_t)dtype;
  drvd_t dd;
  strncpy( dd.port,port,PLEN); dd.port[PLEN-1]=0;
  strncpy( dd.name,name,PLEN); dd.name[PLEN-1]=0;
  dd.slave=slave; dd.addr=addr; dd.len=len; dd.dt=dt;
  pmbus=new drvMBus( dd,msec);
  return(asynSuccess);
}
static const iocshArg confArg0={"port",iocshArgString};
static const iocshArg confArg1={"slave",iocshArgInt};
static const iocshArg confArg2={"addr",iocshArgInt};
static const iocshArg confArg3={"len",iocshArgInt};
static const iocshArg confArg4={"dtype",iocshArgInt};
static const iocshArg confArg5={"name",iocshArgString};
static const iocshArg confArg6={"msec",iocshArgInt};
static const iocshArg* const confArgs[]={&confArg0,&confArg1,&confArg2,
                &confArg3,&confArg4,&confArg5,&confArg6};
static const iocshFuncDef confFuncDef={"drvMBusConfig",7,confArgs};
static void confCallFunc(const iocshArgBuf *args){
  drvMBusConfig( args[0].sval,args[1].ival,args[2].ival,args[3].ival,
                args[4].ival,args[5].sval,args[6].ival);
}

void drvMBusRegister(void){
  iocshRegister(&confFuncDef,confCallFunc);
}
epicsExportRegistrar(drvMBusRegister);
}
